### ISSUE TITLE
Add "additionalProperties" : false to objects by default

### ISSUE THREAD

**Issue by ALEUT** [victools/jsonschema-generator#22]
It would be useful to add `"additionalProperties" : false` for objects (`"type" : "object"`) because by default `additionalProperties` is true and it means, that this object can contain other fields. But if the source of the schema is Java files, it's impossible to have other fields.

Currently I use such a code as workaround:
```java
private static final String TAG_ADDITIONAL_PROPERTIES = "additionalProperties";

.with((jsonSchemaTypeNode, scope, config) -> {
	JsonNode type = jsonSchemaTypeNode.get(TAG_TYPE);
	if (type != null && type.isTextual() && type.textValue().equals(TAG_TYPE_OBJECT)) {
		if (jsonSchemaTypeNode.has(TAG_PROPERTIES)) {
			jsonSchemaTypeNode.put(TAG_ADDITIONAL_PROPERTIES, false);
		}
	}
})
```

Or, as an option, have it configurable using `Option`.

**Comment by CarstenWickner** [comment]
I could also imagine that additional properties are allowed, but will simply be ignored by the Java implementation.

Making this a standard `Option` sounds very reasonable.
You’re welcome to create a PR. Otherwise, I might have a look during the next couple of days myself.

**Comment by ALEUT** [comment]
@CarstenWickner I will be overloaded in nearest week, it would be great it you will look into it.

**Comment by CarstenWickner** [comment]
Hi @ALEUT,

I had a first look into this now and am considering to add this alongside the other fields via its dedicated resolvers in order to allow for things like the Jackson annotation `@JsonIgnoreProperties(ignoreUnknown = true)` to be considered per type instead of just treating this as a single global option. That would still allow you to easily disable it for all types.

However, according to the JSON Schema specification:
> The `additionalProperties` keyword may be either a boolean or an object.
> If `additionalProperties` is an object, that object is a schema that will be used to validate any additional properties not listed in `properties`.

Only catering for the `boolean` side of things seems also rather limited and hard to extend later without a breaking change.
I don't have a nice idea yet how to allow this kind of configuration with type "Boolean | Schema | null"  in a clean but convenient way.

I'm open to suggestions. :)

**Comment by ALEUT** [comment]
Hi @CarstenWickner 
Can it be optional Option then in this case?

**Comment by CarstenWickner** [comment]
Hi @ALEUT,

All `Option`s are optional, but I get what you mean.

However, that wouldn’t solve the issue of having to:
1. potentially reconcile it later with a Schema value and
2. allow this to be set per type and not globally – as per the mentioned Jackson annotation or for special types like a `Map` that could include all kinds of properties but maybe only of a specific type.

**Comment by CarstenWickner** [comment]
I got an idea: the configuration could expect a `Class<?>` to be defined that determines the type of properties (for which to dynamically create a sub-schema).
One special case would be the pseudo-class `java.lang.Void` which could represent the `additionalProperties: false` case.

I'll try that out tonight and see where it leads me.

**Comment by CarstenWickner** [comment]
Hi @ALEUT,

I've created a pull request now with a proposed solution.
The expected usage here for `additionalProperties: false` on all objects would be something like this (depending on what is being treated as `type: "object"` according to other configurations):
```java
SchemaGeneratorConfigBuilder.forTypesInGeneral()
    .withAdditionalPropertiesResolver(scope ->
        scope.isContainerType()
        || scope.getType().isInstanceOf(Number.class)
        || scope.getType().isInstanceOf(String.class)
        || scope.getType().isPrimitive()
            ? null
            : Void.class);
```
E.g. if you would override the handling of container types as `"object"` instead of `"array"` you'd probably also want to see the `"additionalProperties": false` being added.
That's the current challenge I see preventing a global/simple configuraton. The alternative would be very similar to your workaround, but that wouldn't allow for things like the aforementioned Jackson annotation to do this specifically per type.

**Comment by CarstenWickner** [comment]
The basic support for `additionalProperties` has been added now (in #28) but has not yet been released, in order to address @ALEUT's concerns regarding usability.

I might get something done within the next couple of days but will need to see how easy it actually is to achieve what I have in mind.

**Comment by ALEUT** [comment]
@CarstenWickner it's not clear how `"array"` can be handled as `"object"`. Why not just add  `"additionalProperties": false` in case of default `type: "object"` and if any of `"properties"` or `"patternProperties"` are present (Option) and if user needs more detailed configuration - he will just skip this option and will do anything he want in `SchemaGeneratorConfigBuilder with(TypeAttributeOverride override)`?

PS: currently I process the maps in such a way:
```
SchemaGeneratorConfig schemaGeneratorConfig = new SchemaGeneratorConfigBuilder(objectMapper, OptionPreset.PLAIN_JSON)
	.with(Option.DEFINITIONS_FOR_ALL_OBJECTS)
	.with((javaType, context) -> {
		final ObjectNode newDefinition;

		Class<?> inspectedClass = javaType.getErasedType();
		if (inspectedClass == Map.class) {
			List<ResolvedType> mapParameters = javaType.getTypeParameters();
			if (mapParameters.size() == 2) { // key + value
				ResolvedType valueType = mapParameters.get(1);
				Class<?> valueClass = valueType.getErasedType();

				final String keyPattern = "^\\S+$";
				final ObjectNode valueNode = context.createDefinition(valueType);

				newDefinition = objectMapper.createObjectNode();
				newDefinition.put(TAG_TYPE, TAG_TYPE_OBJECT)
					.set(TAG_PATTERN_PROPERTIES, objectMapper.createObjectNode().set(keyPattern, valueNode));
			} else { // if Map is not parameterized
				newDefinition = null;
			}
		} else {
			newDefinition = null;
		}

		return (newDefinition != null) ? new CustomDefinition(newDefinition) : null;
	})
	...
	.build();
```

Output example:
```
"Map(Machine,DurationTime)" : {
      "type" : "object",
      "patternProperties" : {
        "^\\S+$" : {
          "$ref" : "#/definitions/DurationTime"
        }
      },
      "additionalProperties" : false
    },
```

**Comment by CarstenWickner** [comment]
HI @ALEUT, 

I'm afraid I don't fully understand your question:
> it's not clear how "array" can be handled as "object"

Regarding your given example, three points:
1. I've just added support for `patternProperties` as well (since it was almost the same as the `additionalProperties`)..
2. Since you want to set `additionalProperties` to `false` here, your resolver could simply return `Void.class` for a `Map`.
3. To achieve your given output, you could now do the following:
```java
SchemaGeneratorConfigBuilder configBuilder = new SchemaGeneratorConfigBuilder(objectMapper, OptionPreset.PLAIN_JSON);
configBuilder.forTypesInGeneral()
    .withAdditionalPropertiesResolver(scope -> scope.getType().isInstanceOf(Map.class) ? Void.class : null)
    .withPatternPropertiesResolver(scope -> {
        if (scope.getType().isInstanceOf(Map.class)) {
            Map<String, Type> patterns = new HashMap<>();
            patterns.put("^\\S+$", scope.getTypeParameterFor(Map.class, 1));
            return patterns;
        }
        return null;
    });
```
Please beware, the above is not yet released but will be in 4.2.0 then.
What do you reckon?

**Comment by ALEUT** [comment]
@CarstenWickner as you can see there is 2 options in my code for map:
1) map has types (generics) and I can generate patternProperties for it, ex:
```
"Map(Machine,DurationTime)" : {
      "type" : "object",
      "patternProperties" : {
        "^\\S+$" : {
          "$ref" : "#/definitions/DurationTime"
        }
      },
      "additionalProperties" : false
    },
```
2) Map doesn't have generics, I don't know the type of value and the generated code should be like 
```
"Map" : {
      "type" : "object"
    },
```
It's incorrect to place `additionalProperties: false` here because it will mean that any property will be treated as invalid. Also we can't define `patternProperties` because we don't know the value type.

**Comment by ALEUT** [comment]
> I'm afraid I don't fully understand your question:
> 
> > it's not clear how "array" can be handled as "object"

It's related to this one:


> E.g. if you would override the handling of container types as `"object"` instead of `"array"` you'd probably also want to see the `"additionalProperties": false` being added.

**Comment by CarstenWickner** [comment]
Hi @ALEUT,

- One more update on this: I introduced a new `Option` for it now (which internally uses the new resolver configurations).
- Your use-case could now be achieved with the following:
```java
SchemaGeneratorConfigBuilder configBuilder = new SchemaGeneratorConfigBuilder(objectMapper)
    .with(Option.FORBIDDEN_ADDITIONAL_PROPERTIES_BY_DEFAULT);
configBuilder.forTypesInGeneral()
    .withAdditionalPropertiesResolver(scope -> {
        if (scope.getType().isInstanceOf(Map.class)) {
            ResolvedType valueType = scope.getTypeParameterFor(Map.class, 1);
            if (valueType.getErasedType() == Object.class) {
                return Object.class;
            }
        }
        return null;
    })
    .withPatternPropertiesResolver(scope -> {
        if (scope.getType().isInstanceOf(Map.class)) {
            ResolvedType valueType = scope.getTypeParameterFor(Map.class, 1);
            if (valueType.getErasedType() != Object.class) {
                Map<String, Type> patterns = new HashMap<>();
                patterns.put("^\\S+$", valueType);
                return patterns;
            }
        }
        return null;
    });
```
I'm aware it's not really getting shorter, but at least you wouldn't have to mess with generated JSON nodes directly. 😅
- I was also considering somehow including your handling of Maps as another `Option` but I cannot seem to wrap my head around where your non-white-space pattern comes from that you apply to all Maps -- or is that just for your particular `Machine` type that you can make that assumption?
- As to the handling of collections as `object` instead of `array`, I have one [example in the tests](https://github.com/victools/jsonschema-generator/blob/5d0fbf29c1ba71db448d1d21e2197e92a6cea101/src/test/java/com/github/victools/jsonschema/generator/SchemaGeneratorTest.java#L115), achieving that through a custom definition. But that is more for the case when using the generated JSON Schema to document your Java structure and not a JSON being produced from or parsed to it.

**Comment by ALEUT** [comment]
@CarstenWickner cool! One more question: without defining new resolvers and option does it work as before?

**Comment by CarstenWickner** [comment]
Yes, @ALEUT the backward-compatibility is preserved.

**Comment by ALEUT** [comment]
> * I was also considering somehow including your handling of Maps as another `Option` but I cannot seem to wrap my head around where your non-white-space pattern comes from that you apply to all Maps -- or is that just for your particular `Machine` type that you can make that assumption?

It's just limitation for my case - in general case it can be any string.

**Comment by ALEUT** [comment]
Hi @CarstenWickner when do you plan to release it?

**Comment by CarstenWickner** [comment]
Hi @ALEUT,
I plan on doing the release tonight. Just didn’t get to it yesterday.

**Comment by CarstenWickner** [comment]
Hi @ALEUT,

I'm happy to let you know: release [v4.2.0](https://github.com/victools/jsonschema-generator/releases/tag/v4.2.0) has just been published. 🎉

**Comment by ALEUT** [comment]
@CarstenWickner 
My results:
Turned on .with(Option.FORBIDDEN_ADDITIONAL_PROPERTIES_BY_DEFAULT);

1) It generates property for enums:
```
"BlaBla" : {
      "type" : "string",
      "enum" : [ "DEPTH", "WIDTH" ],
      "additionalProperties" : false
    },
```

2) And for refs:
```
"BlaBla" : {
      "$ref" : "#/definitions/_ProcessingUnit",
      "additionalProperties" : false
    },
```

3) For `string`:
```
"Date" : {
      "type" : "string",
      "additionalProperties" : false
    },
```

4) For `anyOf`

Not sure if it's incorrect, but I see no difference with absence of the property (if I will try to write anything else except value in enum in case 1 (e.g. `DEPTH123`), I will receive an error on validation, the same for case 2).

**Comment by ALEUT** [comment]
@CarstenWickner why do we need check `if (valueType.getErasedType() != Object.class) {` in `withPatternPropertiesResolver`?

**Comment by CarstenWickner** [comment]
HI @ALEUT,

1. Can you maybe share a complete configuration example where the above can be reproduced?
I specifically excluded the `additionalProperties` in combination with the standard "simple types", which is why I'm surprised to see it for strings (your case 3). For the simplified enums (your case 1), I guess I must have forgotten the specific handling.
2. Regarding case 2, I must have messed something up, since that only explicitly allowed starting with Draft 2019-09 and I haven't added that official support yet. Under Draft 7 anything besides `$ref` is supposed to be ignored anyway.
3. Since you didn't provide an example for point 4, I'm guessing it's the same issue as with point 2. Really strange.

As to the check for `Object.class` in the `patternProperties` handling: I needed to avoid always adding `{ "type": "object" }` in case of `Object.class` being provided, as that might just as well stand for a `{ "type": "string" }` or `{ "type": "number" }` which would then fail a validation.
That's why `Object.class` in this context is treated as complete wild-card. Might make sense to do this globally as well to be on the safe side. Then, the special handling for `patternProperties` wouldn't be necessary at all.

**Comment by ALEUT** [comment]
Hi @CarstenWickner 
1) Can't provide it now - it's from my project.

For case 3 false alarm - I added this type manually: 
```
if (inspectedClass == java.sql.Date.class) {
					newDefinition = objectMapper.createObjectNode()
						.put(TAG_TYPE, TAG_TYPE_STRING);
				}
```
But for some reason additionalProperties was added later.

For case 4: I also added this anyOf manually.

**Comment by CarstenWickner** [comment]
It seems that’s then only in combination with custom definitions. This particular `Option` should probably rather be ignored then. I‘ll have to somehow expose to the resolvers then whether the targeted type is a custom one or not. Unfortunately not straightforward to remember and pass-through.

Not sure when I will get to it, but that just means additional care needs to be taken with custom definitions, since all other configurations still apply to them.

**Comment by CarstenWickner** [comment]
Hi @ALEUT,
In regards to your custom definition for `java.sql.Date`: do you have more standard types for which you add these kinds of custom definitions?
I‘m happy to extend the list of standard types that are supposed to be treated as anything but “object”.

**Comment by ALEUT** [comment]
Hi @CarstenWickner 
We have custom serializers and deserializers, that's why I treat `java.sql.Date` as string in my case. It can't be applicable in generic case.

**Comment by CarstenWickner** [comment]
Hi @ALEUT,

I've made some adjustments (in #34) for the handling of the general `Object.class` and also when type-level attributes (such as `"additionalProperties"` through `Option.FORBIDDEN_ADDITIONAL_PROPERTIES_BY_DEFAULT`) are being added.
Now, it should no longer be added to your custom non-`object` definitions anymore. In case of your manually added `"anyOf"` you may still have to take of it yourself though.

This has been released as [v4.3.0](https://github.com/victools/jsonschema-generator/releases/tag/v4.3.0) and should hopefully conclude the `"additionalProperties"` implementation now.
